src/main/java/io/confluent/connect/elasticsearch/OffsetTracker.java

```diff
@@ -0,0 +1,151 @@
+      this.offset = offset;
+    }
+
+    /**
+     * Marks the offset as processed (ready to report to preCommit)
+     */
+    public void markProcessed() {
+      processed = true;
+    }
+
+    public boolean isProcessed() {
+      return processed;
+    }
+  }
+
+  /**
+   * Partitions are no longer owned, we should release all related resources.
+   */
+  public synchronized void closePartitions(Collection<TopicPartition> topicPartitions) {
+    topicPartitions.forEach(tp -> {
+      Map<Long, OffsetState> offsets = offsetsByPartition.remove(tp);
+      if (offsets != null) {
+        numEntries.getAndAdd(offsets.size());
+      }
+      maxOffsetByPartition.remove(tp);
+    });
+  }
+
+  /**
+   * This method assumes that new records are added in offset order.
+   * Older records can be re-added, and the same Offset object will be return if its
+   * offset hasn't been reported yet.
+   */
+  public synchronized OffsetState addPendingRecord(SinkRecord sinkRecord) {
+    log.trace("Adding pending record");
+    TopicPartition tp = new TopicPartition(sinkRecord.topic(), sinkRecord.kafkaPartition());
+    Long partitionMax = maxOffsetByPartition.get(tp);
+    if (partitionMax == null || sinkRecord.kafkaOffset() > partitionMax) {
+      numEntries.incrementAndGet();
+      return offsetsByPartition
+              // Insertion order needs to be maintained
+              .computeIfAbsent(tp, key -> new LinkedHashMap<>())
+              .computeIfAbsent(sinkRecord.kafkaOffset(), OffsetState::new);
+    } else {
+      return new OffsetState(sinkRecord.kafkaOffset());
+    }
+  }
+
+  /**
+   * @return overall number of entries
+   */
+  public long numEntries() {
+    return numEntries.get();
+  }
+
+  /**
+   * Move offsets to the highest we can.
+   */
+  public synchronized void updateOffsets() {
+    log.trace("Updating offsets");
+    offsetsByPartition.forEach(((topicPartition, offsets) -> {
+      Long max = maxOffsetByPartition.get(topicPartition);
+      boolean newMaxFound = false;
+      Iterator<OffsetState> iterator = offsets.values().iterator();
+      while (iterator.hasNext()) {
+        OffsetState offsetState = iterator.next();
+        if (offsetState.isProcessed()) {
+          iterator.remove();
+          numEntries.decrementAndGet();
+          if (max == null || offsetState.offset > max) {
+            max = offsetState.offset;
+            newMaxFound = true;
+          }
+        } else {
+          break;
+        }
+      }
+      if (newMaxFound) {
+        maxOffsetByPartition.put(topicPartition, max);
+      }
+    }));
+    log.trace("Updated offsets, num entries: {}", numEntries);
+  }
+
+  /**
+   * @return offsets to commit
+   */
+  public synchronized Map<TopicPartition, OffsetAndMetadata> offsets() {
+    return maxOffsetByPartition.entrySet().stream()
+            .collect(toMap(
+                Map.Entry::getKey,
+                e -> new OffsetAndMetadata(e.getValue() + 1)));
```

Comment:
Add comment or mention in JavaDoc why "+1"?

Reply:
Added